core/embed/trezorhal/optiga/optiga.c

```diff
@@ -109,3 +158,432 @@ bool optiga_random_buffer(uint8_t *dest, size_t size) {
+
+    // Combine intermediate result with OID_PIN_CMAC.
+    optiga_result res =
+        optiga_encrypt_sym(OPTIGA_SYM_MODE_CMAC, OID_PIN_CMAC, result,
+                           sizeof(result), buffer, sizeof(buffer), &size);
+    if (res != OPTIGA_SUCCESS) {
+      memzero(buffer, sizeof(buffer));
+      memzero(result, sizeof(result));
+      memzero(&ctx, sizeof(ctx));
+      return false;
+    }
+
+    hmac_sha256_Update(&ctx, buffer, size);
+
+    ui_progress(PIN_DERIVE_SECS * 1000 * step / total_steps);
+    step++;
+
+    // Combine intermediate result with OID_PIN_HMAC
+    res = optiga_encrypt_sym(OPTIGA_SYM_MODE_HMAC_SHA256, OID_PIN_HMAC, result,
+                             sizeof(result), buffer, sizeof(buffer), &size);
+    if (res != OPTIGA_SUCCESS) {
+      memzero(buffer, sizeof(buffer));
+      memzero(result, sizeof(result));
+      memzero(&ctx, sizeof(ctx));
+      return false;
+    }
+
+    hmac_sha256_Update(&ctx, buffer, size);
+
+    ui_progress(PIN_DERIVE_SECS * 1000 * step / total_steps);
+    step++;
+
+    // Combine intermediate result with OID_PIN_ECDH
+    uint8_t encoded_point[BIT_STRING_HEADER_SIZE + 65] = {0x03, 0x42, 0x00};
+    /* TODO
+    if(!hash_to_curve_optiga(result, uint8_t
+    &encoded_point[BIT_STRING_HEADER_SIZE])) { memzero(buffer, sizeof(buffer));
+      memzero(result, sizeof(result));
+      memzero(&ctx, sizeof(ctx));
+      return false;
+    }
+    */
+    res =
+        optiga_calc_ssec(OPTIGA_CURVE_P256, OID_PIN_ECDH, encoded_point,
+                         sizeof(encoded_point), buffer, sizeof(buffer), &size);
+    memzero(encoded_point, sizeof(encoded_point));
+    if (res != OPTIGA_SUCCESS) {
+      memzero(buffer, sizeof(buffer));
+      memzero(result, sizeof(result));
+      memzero(&ctx, sizeof(ctx));
+      return false;
+    }
+
+    hmac_sha256_Update(&ctx, buffer, size);
+
+    ui_progress(PIN_DERIVE_SECS * 1000 * step / total_steps);
+    step++;
+
+    hmac_sha256_Final(&ctx, result);
+  }
+
+  memcpy(secret, result, sizeof(result));
+  memzero(buffer, sizeof(buffer));
+  memzero(result, sizeof(result));
+  memzero(&ctx, sizeof(ctx));
+  return true;
+}
+
+bool optiga_pin_set(OPTIGA_UI_PROGRESS ui_progress, const uint8_t *pin,
+                    size_t pin_size,
+                    const uint8_t storage_salt[STORAGE_SALT_SIZE],
+                    uint8_t secret[OPTIGA_PIN_SECRET_SIZE]) {
+  if (!optiga_pin_init_metadata() || !optiga_pin_init_stretch()) {
+    return false;
+  }
+
+  // Combine the PIN with salt from the flash storage before sending it to the
+  // Optiga. This ensures that in the unlikely case of an attacker recording
+  // communication between the MCU and Optiga, they will not gain knowledge of
+  // the PIN without at least having to conduct a brute-force search.
+  uint8_t salted_pin[OPTIGA_PIN_SECRET_SIZE] = {0};
+  hmac_sha256(pin, pin_size, storage_salt, STORAGE_SALT_SIZE, salted_pin);
```

Comment:
The time complexity of the attack is `2^(8 * STORAGE_SALT_SIZE + pin_entropy) = 2^(32 + pin_entropy)`, measured in the number of `HMAC-SHA256` computations.

The purpose of the secure element is to allow a user to safely use a short pin. However, since the salt has only 32 bits, the brute-force attack is feasible for short pins.

Reply:
My assumption here is that an attacker that has physical access to the device knows the storage salt.

Reply:
I don't understand what you mean. If you assume that the attacker knows the storage salt, then there is no reason to mix the salt with the pin.

Reply:
This thought occurred to me: If we want a Trezor with a malicious Optiga to be as secure as a Trezor without Optiga, then we have to replace the [HMAC](https://github.com/trezor/trezor-firmware/blob/a19732de622892fe5ff359f934203731d57b1af7/core/embed/trezorhal/optiga/optiga.c#L452) with [PBKDF](https://github.com/trezor/trezor-firmware/blob/a19732de622892fe5ff359f934203731d57b1af7/storage/storage.c#L489-L523). Otherwise, it is easier for the attacker to bruteforce the salted pin sent to the optiga than to bruteforce the pin verification code in the storage.

Reply:
The same applies to [this](https://github.com/trezor/trezor-firmware/blob/a19732de622892fe5ff359f934203731d57b1af7/core/embed/trezorhal/optiga/optiga.c#L520) HMAC.